File: apps/api/AssetManagement.Api/Controllers/SettingsController.cs
using System.Security.Claims;
using AssetManagement.Api.Data;
using AssetManagement.Api.DTOs;
using AssetManagement.Api.Models;
using AssetManagement.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace AssetManagement.Api.Controllers;

[ApiController]
[Authorize]
[Route("api/v1/[controller]")]
public class SettingsController(AppDbContext db, IAuditService audit) : ControllerBase
{
    private string GetUserName() => User.FindFirst(ClaimTypes.Name)?.Value ?? "Unknown";
    private Guid? GetUserId()
    {
        var claim = User.FindFirst(ClaimTypes.NameIdentifier)?.Value;
        return claim is not null && Guid.TryParse(claim, out var id) ? id : null;
    }

    private bool IsAdmin() => User.IsInRole("Admin");

    private async Task<string> GetSetting(string key, string defaultValue = "")
    {
        var setting = await db.SystemSettings.FindAsync(key);
        return setting?.Value ?? defaultValue;
    }

    private async Task SetSetting(string key, string value, string updatedBy)
    {
        var setting = await db.SystemSettings.FindAsync(key);
        if (setting is null)
        {
            db.SystemSettings.Add(new SystemSetting
            {
                Key = key,
                Value = value,
                UpdatedBy = updatedBy
            });
        }
        else
        {
            setting.Value = value;
            setting.UpdatedAt = DateTime.UtcNow;
            setting.UpdatedBy = updatedBy;
        }
    }

    [HttpGet("system")]
    public async Task<ActionResult<SystemSettingsDto>> GetSystem()
    {
        var dto = new SystemSettingsDto(
            await GetSetting("org.name", "My Organisation"),
            await GetSetting("org.currency", "GBP"),
            await GetSetting("org.dateFormat", "DD/MM/YYYY"),
            int.TryParse(await GetSetting("org.defaultPageSize", "25"), out var ps) ? ps : 25
        );
        return Ok(dto);
    }

    [HttpPut("system")]
    public async Task<ActionResult<SystemSettingsDto>> UpdateSystem(SystemSettingsDto request)
    {
        if (!IsAdmin()) return Forbid();

        var userName = GetUserName();
        await SetSetting("org.name", request.OrgName, userName);
        await SetSetting("org.currency", request.Currency, userName);
        await SetSetting("org.dateFormat", request.DateFormat, userName);
        await SetSetting("org.defaultPageSize", request.DefaultPageSize.ToString(), userName);
        await db.SaveChangesAsync();

        await audit.LogAsync(new AuditEntry(
            "Updated", "SystemSettings", "system", "System Settings",
            "System settings updated", GetUserId(), userName));

        return Ok(request);
    }

    [HttpGet("alerts")]
    public async Task<ActionResult<AlertSettingsDto>> GetAlerts()
    {
        if (!IsAdmin()) return Forbid();

        var dto = new AlertSettingsDto(
            (await GetSetting("alerts.warranty.enabled", "true")) == "true",
            (await GetSetting("alerts.certificate.enabled", "true")) == "true",
            (await GetSetting("alerts.licence.enabled", "true")) == "true",
            await GetSetting("alerts.thresholds", "90,30,14,7"),
            await GetSetting("alerts.smtp.host"),
            int.TryParse(await GetSetting("alerts.smtp.port", "587"), out var port) ? port : 587,
            await GetSetting("alerts.smtp.username"),
            await GetSetting("alerts.smtp.password"),
            await GetSetting("alerts.smtp.fromAddress"),
            await GetSetting("alerts.slack.webhookUrl"),
            await GetSetting("alerts.recipients")
        );
        return Ok(dto);
    }

    [HttpPut("alerts")]
    public async Task<ActionResult<AlertSettingsDto>> UpdateAlerts(AlertSettingsDto request)
    {
        if (!IsAdmin()) return Forbid();

        var userName = GetUserName();
        await SetSetting("alerts.warranty.enabled", request.WarrantyEnabled.ToString().ToLower(), userName);
        await SetSetting("alerts.certificate.enabled", request.CertificateEnabled.ToString().ToLower(), userName);
        await SetSetting("alerts.licence.enabled", request.LicenceEnabled.ToString().ToLower(), userName);
        await SetSetting("alerts.thresholds", request.Thresholds, userName);
        await SetSetting("alerts.smtp.host", request.SmtpHost, userName);
        await SetSetting("alerts.smtp.port", request.SmtpPort.ToString(), userName);
        await SetSetting("alerts.smtp.username", request.SmtpUsername, userName);
        await SetSetting("alerts.smtp.password", request.SmtpPassword, userName);
        await SetSetting("alerts.smtp.fromAddress", request.SmtpFromAddress, userName);
        await SetSetting("alerts.slack.webhookUrl", request.SlackWebhookUrl, userName);
        await SetSetting("alerts.recipients", request.Recipients, userName);
        await db.SaveChangesAsync();

        await audit.LogAsync(new AuditEntry(
            "Updated", "AlertSettings", "alerts", "Alert Settings",
            "Alert settings updated", GetUserId(), userName));

        return Ok(request);
    }
}

File: apps/api-kt/src/main/kotlin/com/assetmanagement/api/controller/SettingsController.kt
package com.assetmanagement.api.controller

import com.assetmanagement.api.dto.AlertSettingsDto
import com.assetmanagement.api.dto.SystemSettingsDto
import com.assetmanagement.api.model.SystemSetting
import com.assetmanagement.api.repository.SystemSettingRepository
import com.assetmanagement.api.service.AlertSchedulerService
import com.assetmanagement.api.service.AuditEntry
import com.assetmanagement.api.service.AuditService
import com.assetmanagement.api.service.CurrentUserService
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.web.bind.annotation.*
import java.time.Instant

@RestController
@RequestMapping("/api/v1/settings")
class SettingsController(
    private val systemSettingRepository: SystemSettingRepository,
    private val auditService: AuditService,
    private val currentUserService: CurrentUserService,
    private val alertSchedulerService: AlertSchedulerService
) {
    private val log = LoggerFactory.getLogger(SettingsController::class.java)
    private val secretMask = "********"

    private fun maskSecret(value: String): String =
        if (value.isNotEmpty()) secretMask else ""

    private fun maskWebhookUrl(value: String): String =
        if (value.length > 20) value.substring(0, 20) + "..." else value

    private fun isAdmin(): Boolean =
        SecurityContextHolder.getContext().authentication?.authorities?.any { it.authority == "ROLE_Admin" } == true

    private fun getSetting(key: String, default_: String = ""): String =
        systemSettingRepository.findByKey(key)?.value ?: default_

    private fun setSetting(key: String, value: String, updatedBy: String) {
        val setting = systemSettingRepository.findByKey(key)
        if (setting != null) {
            setting.value = value
            setting.updatedAt = Instant.now()
            setting.updatedBy = updatedBy
            systemSettingRepository.save(setting)
        } else {
            systemSettingRepository.save(SystemSetting(key = key, value = value, updatedBy = updatedBy))
        }
    }

    @GetMapping("/system")
    fun getSystem(): ResponseEntity<SystemSettingsDto> {
        val dto = SystemSettingsDto(
            orgName = getSetting("org.name", "My Organisation"),
            currency = getSetting("org.currency", "GBP"),
            dateFormat = getSetting("org.dateFormat", "DD/MM/YYYY"),
            defaultPageSize = getSetting("org.defaultPageSize", "25").toIntOrNull() ?: 25
        )
        return ResponseEntity.ok(dto)
    }

    @PutMapping("/system")
    fun updateSystem(@RequestBody request: SystemSettingsDto): ResponseEntity<Any> {
        if (!isAdmin()) return ResponseEntity.status(403).build()

        val userName = currentUserService.userName
        setSetting("org.name", request.orgName, userName)
        setSetting("org.currency", request.currency, userName)
        setSetting("org.dateFormat", request.dateFormat, userName)
        setSetting("org.defaultPageSize", request.defaultPageSize.toString(), userName)

        auditService.log(AuditEntry("Updated", "SystemSettings", "system", "System Settings",
            "System settings updated", currentUserService.userId, userName))

        return ResponseEntity.ok(request)
    }

    @GetMapping("/alerts")
    fun getAlerts(): ResponseEntity<Any> {
        if (!isAdmin()) return ResponseEntity.status(403).build()

        val dto = AlertSettingsDto(
            warrantyEnabled = getSetting("alerts.warranty.enabled", "true") == "true",
            certificateEnabled = getSetting("alerts.certificate.enabled", "true") == "true",
            licenceEnabled = getSetting("alerts.licence.enabled", "true") == "true",
            thresholds = getSetting("alerts.thresholds", "90,30,14,7"),
            emailProvider = getSetting("alerts.email.provider", "smtp"),
            smtpHost = getSetting("alerts.smtp.host"),
            smtpPort = getSetting("alerts.smtp.port", "587").toIntOrNull() ?: 587,
            smtpUsername = getSetting("alerts.smtp.username"),
            smtpPassword = maskSecret(getSetting("alerts.smtp.password")),
            smtpFromAddress = getSetting("alerts.smtp.fromAddress"),
            graphTenantId = getSetting("alerts.graph.tenantId"),
            graphClientId = getSetting("alerts.graph.clientId"),
            graphClientSecret = maskSecret(getSetting("alerts.graph.clientSecret")),
            graphFromAddress = getSetting("alerts.graph.fromAddress"),
            slackWebhookUrl = maskWebhookUrl(getSetting("alerts.slack.webhookUrl")),
            recipients = getSetting("alerts.recipients"),
            scheduleType = getSetting("alerts.schedule.type", "disabled"),
            scheduleTime = getSetting("alerts.schedule.time", "09:00"),
            scheduleDay = getSetting("alerts.schedule.day", "MONDAY")
        )
        return ResponseEntity.ok(dto)
    }

    @PutMapping("/alerts")
    fun updateAlerts(@RequestBody request: AlertSettingsDto): ResponseEntity<Any> {
        if (!isAdmin()) return ResponseEntity.status(403).build()

        val userName = currentUserService.userName
        setSetting("alerts.warranty.enabled", request.warrantyEnabled.toString().lowercase(), userName)
        setSetting("alerts.certificate.enabled", request.certificateEnabled.toString().lowercase(), userName)
        setSetting("alerts.licence.enabled", request.licenceEnabled.toString().lowercase(), userName)
        setSetting("alerts.thresholds", request.thresholds, userName)
        setSetting("alerts.email.provider", request.emailProvider, userName)
        setSetting("alerts.smtp.host", request.smtpHost, userName)
        setSetting("alerts.smtp.port", request.smtpPort.toString(), userName)
        setSetting("alerts.smtp.username", request.smtpUsername, userName)
        if (request.smtpPassword != secretMask) setSetting("alerts.smtp.password", request.smtpPassword, userName)
        setSetting("alerts.smtp.fromAddress", request.smtpFromAddress, userName)
        setSetting("alerts.graph.tenantId", request.graphTenantId, userName)
        setSetting("alerts.graph.clientId", request.graphClientId, userName)
        if (request.graphClientSecret != secretMask) setSetting("alerts.graph.clientSecret", request.graphClientSecret, userName)
        setSetting("alerts.graph.fromAddress", request.graphFromAddress, userName)
        if (!request.slackWebhookUrl.endsWith("...")) setSetting("alerts.slack.webhookUrl", request.slackWebhookUrl, userName)
        setSetting("alerts.recipients", request.recipients, userName)
        setSetting("alerts.schedule.type", request.scheduleType, userName)
        setSetting("alerts.schedule.time", request.scheduleTime, userName)
        setSetting("alerts.schedule.day", request.scheduleDay, userName)

        auditService.log(AuditEntry("Updated", "AlertSettings", "alerts", "Alert Settings",
            "Alert settings updated", currentUserService.userId, userName))

        try {
            alertSchedulerService.reschedule()
        } catch (e: Exception) {
            log.error("Failed to reschedule alerts", e)
        }

        return ResponseEntity.ok(request)
    }
}
